moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/detail/ompl_constraints.h

```diff
@@ -0,0 +1,325 @@
+
+#include <ompl/base/Constraint.h>
+
+#include <moveit/ompl_interface/detail/threadsafe_state_storage.h>
+#include <moveit/robot_model/robot_model.h>
+#include <moveit_msgs/Constraints.h>
+#include <moveit/macros/class_forward.h>
+
+namespace ompl_interface
+{
+MOVEIT_CLASS_FORWARD(BaseConstraint);
+MOVEIT_CLASS_FORWARD(BoxConstraint);
+MOVEIT_CLASS_FORWARD(OrientationConstraint);
+
+/** \brief Represents upper and lower bound on a scalar value (double).
+ *
+ * OMPL ConstrainedStateSpace requires a model of the constraints given as:
+ *  f1(joint_values) = 0
+ *  f2(joint_values) = 0
+ *  f3(joint_values) = 0
+ *  ...
+ *
+ * So we use a penalty function to convert bounds to an equality constraint.
+ * If you do need equality constraint, you can represent them by setting the upper bound and lower bound almost equal.
+ * Or you can use the EqualityPositionConstraint version by setting the name of the constraint to
+ * "use_equality_constraints". But the latter will ignore bounds on other dimensions.
+ * **/
+struct Bounds
+{
+  double lower, upper;
+
+  /** \brief Distance to region inside bounds
+   *
+   * Distance of a given value outside the bounds, zero inside the bounds.
+   * Creates a penalty function that looks like this:
+   *
+   * (penalty) ^
+   *           | \         /
+   *           |  \       /
+   *           |   \_____/
+   *           |----------------> (variable to be constrained)
+   *
+   * TODO(jeroendm) Change it to a penalty function that has a continuous derivative, so we can use AtlasStateSpace and
+   * TangentBundleStateSpace.
+   * */
+  inline double penalty(double value) const;
+
+  /** \brief Derivative of the penalty function
+   * ^
+   * |
+   * | -1-1-1 0 0 0 +1+1+1
+   * |------------------------>
+   * **/
+  inline double derivative(double value) const;
+};
+
+/** \brief Pretty printing of bounds. **/
+std::ostream& operator<<(std::ostream& os, const ompl_interface::Bounds& bound);
+
+/****************************
+ * Base class for constraints
+ * **************************/
+/** \brief Abstract base class for different types of constraints, implementations of ompl::base::Constraint
+ *
+ * To create a constrained state space in OMPL, we need a model of the constraints, that can be written in the form of
+ * equality constraints F(joint_values) = 0. This class uses `Bounds` defined above, to convert:
+ *
+ *    lower_bound < scalar value < upper bound
+ *
+ * into an equation of the form f(x) = 0.
+ *
+ * The 'scalar value' can be the difference between the position or orientation of a link and a target position or
+ * orientation, or any other error metric that can be calculated using the `moveit::core::RobotModel` and
+ * `moveit::core::JointModelGroup`.
+ * */
+class BaseConstraint : public ompl::base::Constraint
+{
+public:
+  BaseConstraint(const robot_model::RobotModelConstPtr& robot_model, const std::string& group,
+                 const unsigned int num_dofs, const unsigned int num_cons_ = 3);
```

Comment:
Also, remove the trailing underscore since this is typically used for member variables.

Reply:
I've added a comment. All constraints currently implemented have 3 constraint equations, so it might just as well be the default. The MoveIt constraint messages such as position and orientation constraints also have 3 dimensions.